compiler/frontend/src/org/jetbrains/kotlin/types/expressions/LabelResolver.kt

```diff
@@ -160,11 +176,28 @@ object LabelResolver {
                 return LabeledReceiverResolutionResult.labelResolutionSuccess(thisReceiver)
             }
             0 -> {
-                val element = resolveNamedLabel(labelName, targetLabel, context.trace)
-                val declarationDescriptor = context.trace.bindingContext[BindingContext.DECLARATION_TO_DESCRIPTOR, element]
+                val element = resolveNamedLabel(
+                    labelName, targetLabel, context.trace,
+                    classNameLabelsEnabled = expression is KtThisExpression
+                            && context.languageVersionSettings.supportsFeature(LanguageFeature.ContextReceivers)
+                )
+                val declarationDescriptor = context.trace.bindingContext[DECLARATION_TO_DESCRIPTOR, element]
                 if (declarationDescriptor is FunctionDescriptor) {
-                    val thisReceiver = declarationDescriptor.extensionReceiverParameter
-                    if (thisReceiver != null) {
+                    val receiverToLabelMap = context.trace.bindingContext[
+                            DESCRIPTOR_TO_NAMED_RECEIVERS,
+                            if (declarationDescriptor is PropertyAccessorDescriptor) declarationDescriptor.correspondingProperty else declarationDescriptor
+                    ]
+                    val thisReceivers = receiverToLabelMap?.entries?.filter {
```

Comment:
It looks kind of strange that this map is only used to look through the values, but not the keys
I'd invert the map, so it would point from label names to the relevant descriptor

Reply:
Thank you, this code looks indeed weird. Fixed in [the commit](LabelNameToReceiverStorage).
I created the separate class `LabelNameToReceiverStorage`, which holds a mapping from label name to a list of receiver descriptors. We need to hold a list (not a single receiver descriptor) because there can be multiple ones. For example:

```kotlin
context(A<Int>, A<String>)
fun f() {
    this@A // We need to report an ambiguity at the call site
}
```